### PR TITLE
Error if id_loan exists and has any duplicated value

### LINKED ISSUE #349: Does match_name() need to error with duplicated `id_loan`?
@georgeharris2deg suggested that the error we throw in `prioritize()` should be thrown also earlier in `match_name()`. 

@georgeharris2deg please correctly me if I missunderstood.

@jdhoffa what do you think?

``` r
library(r2dii.match)
devtools::load_all()
#> Loading r2dii.match

bad <- c(1, 1)
lbk <- tibble::tibble(
  id_loan = bad,
  sector_classification_system = "NACE",
  id_ultimate_parent = "UP15",
  name_ultimate_parent = "Alpine Knits India Pvt. Limited",
  id_direct_loantaker = "C294",
  name_direct_loantaker = "Yuamen Xinneng Thermal Power Co Ltd",
  sector_classification_direct_loantaker = 3511
)

ald <- data.frame(
  name_company = "alpine knits india pvt. limited", 
  sector = "power", 
  alias_ald = "alpineknitsindiapvt ltd"
)

# FIXME: Do we need to throw an error here?
expect_error(
  out <- match_name(lbk, ald)
)
#> Error: `out <- match_name(lbk, ald)` did not throw the expected error.
out
#> # A tibble: 2 x 16
#>   id_loan sector_classifi… id_ultimate_par… name_ultimate_p… id_direct_loant…
#>     <dbl> <chr>            <chr>            <chr>            <chr>           
#> 1       1 NACE             UP15             Alpine Knits In… C294            
#> 2       1 NACE             UP15             Alpine Knits In… C294            
#> # … with 11 more variables: name_direct_loantaker <chr>,
#> #   sector_classification_direct_loantaker <dbl>, id_2dii <chr>, level <chr>,
#> #   sector <chr>, sector_ald <chr>, name <chr>, name_ald <chr>, score <dbl>,
#> #   source <chr>, borderline <lgl>

# Here we do throw an error, but is it too late?
expect_error(prioritize(out))
```

<sup>Created on 2021-02-04 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
<details>
<summary>
Session info
</summary>

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.3 (2020-10-10)
#>  os       Ubuntu 18.04.5 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language en_US:en                    
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/Matamoros           
#>  date     2021-02-04                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  ! package     * version    date       lib source                            
#>    assertthat    0.2.1      2019-03-21 [1] CRAN (R 4.0.3)                    
#>    callr         3.5.1      2020-10-13 [1] CRAN (R 4.0.3)                    
#>    cli           2.2.0      2020-11-20 [1] CRAN (R 4.0.3)                    
#>    crayon        1.3.4      2017-09-16 [1] CRAN (R 4.0.3)                    
#>    data.table    1.13.6     2020-12-30 [1] RSPM (R 4.0.3)                    
#>    DBI           1.1.1      2021-01-15 [1] RSPM (R 4.0.3)                    
#>    desc          1.2.0      2018-05-01 [1] CRAN (R 4.0.0)                    
#>    devtools      2.3.2      2020-09-18 [1] RSPM (R 4.0.2)                    
#>    digest        0.6.27     2020-10-24 [1] CRAN (R 4.0.3)                    
#>    dplyr         1.0.3      2021-01-15 [1] RSPM (R 4.0.3)                    
#>    ellipsis      0.3.1      2020-05-15 [1] CRAN (R 4.0.3)                    
#>    evaluate      0.14       2019-05-28 [1] CRAN (R 4.0.0)                    
#>    fansi         0.4.2      2021-01-15 [1] RSPM (R 4.0.3)                    
#>    fs            1.5.0      2020-07-31 [1] RSPM (R 4.0.2)                    
#>    generics      0.1.0      2020-10-31 [1] CRAN (R 4.0.3)                    
#>    glue          1.4.2      2020-08-27 [1] CRAN (R 4.0.3)                    
#>    highr         0.8        2019-03-20 [1] CRAN (R 4.0.0)                    
#>    htmltools     0.5.1.9000 2021-01-23 [1] Github (rstudio/htmltools@e7f0393)
#>    knitr         1.30       2020-09-22 [1] RSPM (R 4.0.2)                    
#>    lifecycle     0.2.0      2020-03-06 [1] CRAN (R 4.0.3)                    
#>    magrittr      2.0.1      2020-11-17 [1] CRAN (R 4.0.3)                    
#>    memoise       1.1.0      2017-04-21 [1] CRAN (R 4.0.0)                    
#>    pillar        1.4.7      2020-11-20 [1] CRAN (R 4.0.3)                    
#>    pkgbuild      1.2.0      2020-12-15 [1] RSPM (R 4.0.3)                    
#>    pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.0.3)                    
#>    pkgload       1.1.0      2020-05-29 [1] CRAN (R 4.0.0)                    
#>    prettyunits   1.1.1      2020-01-24 [1] CRAN (R 4.0.0)                    
#>    processx      3.4.5      2020-11-30 [1] RSPM (R 4.0.3)                    
#>    ps            1.5.0      2020-12-05 [1] RSPM (R 4.0.3)                    
#>    purrr         0.3.4      2020-04-17 [1] CRAN (R 4.0.0)                    
#>    r2dii.data    0.1.6.9000 2021-02-03 [1] local                             
#>  P r2dii.match * 0.0.8.9000 2021-01-21 [?] local                             
#>    R6            2.5.0      2020-10-28 [1] RSPM (R 4.0.3)                    
#>    remotes       2.2.0      2020-07-21 [1] RSPM (R 4.0.3)                    
#>    rlang         0.4.10     2020-12-30 [1] RSPM (R 4.0.3)                    
#>    rmarkdown     2.6        2020-12-14 [1] RSPM (R 4.0.3)                    
#>    rprojroot     2.0.2      2020-11-15 [1] RSPM (R 4.0.3)                    
#>    rstudioapi    0.13       2020-11-12 [1] RSPM (R 4.0.3)                    
#>    sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 4.0.0)                    
#>    stringdist    0.9.6.3    2020-10-09 [1] CRAN (R 4.0.3)                    
#>    stringi       1.5.3      2020-09-09 [1] RSPM (R 4.0.2)                    
#>    stringr       1.4.0      2019-02-10 [1] CRAN (R 4.0.0)                    
#>    testthat    * 3.0.1      2020-12-17 [1] RSPM (R 4.0.3)                    
#>    tibble        3.0.5      2021-01-15 [1] RSPM (R 4.0.3)                    
#>    tidyr         1.1.2      2020-08-27 [1] RSPM (R 4.0.2)                    
#>    tidyselect    1.1.0      2020-05-11 [1] CRAN (R 4.0.0)                    
#>    usethis       2.0.0.9000 2020-12-11 [1] Github (r-lib/usethis@f96bf2e)    
#>    utf8          1.1.4      2018-05-24 [1] CRAN (R 4.0.3)                    
#>    vctrs         0.3.6      2020-12-17 [1] RSPM (R 4.0.3)                    
#>    withr         2.4.0      2021-01-16 [1] RSPM (R 4.0.3)                    
#>    xfun          0.20       2021-01-06 [1] RSPM (R 4.0.3)                    
#>    yaml          2.2.1      2020-02-01 [1] RSPM (R 4.0.0)                    
#> 
#> [1] /home/mauro/R/x86_64-pc-linux-gnu-library/4.0
#> [2] /usr/local/lib/R/site-library
#> [3] /usr/lib/R/site-library
#> [4] /usr/lib/R/library
#> 
#>  P ── Loaded and on-disk path mismatch.
```

</details>

### PR DESCRIPTION
**Pull request by maurolepore**
Closes #349

Thanks @georgeharris2deg

This PR makes `match_name()` fail if two things happen:

* The input loanbook has a column named `id_loan`.
* The column `id_loan` has at least one duplicated value.

@jdhoffa, notice that the loanbook may not have a column `id_loan` (should we enforce it?).

### CODE REVIEW

File: tests/testthat/_snaps/match_name.md [hunk 1]
```diff
@@ -195,3 +195,8 @@
       ]
     }
 
+# errors if any id_loan is duplicated (#349)
+
+      All values of `id_loan` in a `loanbook` must be unique.
+      Do you need to remove duplicated values (e.g. `1`)?
```
Comment (maurolepore) [inline tests/testthat/_snaps/match_name.md:201]: Here is what the users would see.

File: DESCRIPTION [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Package: r2dii.match
 Title: Tools to Match Financial Portfolios with Climate Data
-Version: 0.0.8.9000
+Version: 0.0.8.9001
```
Comment (maurolepore) [inline DESCRIPTION:3]:
```r
remotes::install_github("2DegreesInvesting/r2dii.match@349_match-name-should-fail-earlier")
# or
remotes::install_github("2DegreesInvesting/r2dii.match@0.0.8.9001")
```

**Comment by georgeharris2deg** [comment]
@maurolepore 
Thanks a lot for this 
If it isn't too late I would like to suggest editing the error message slightly 
**how it is now** 
"All values of `id_loan` in a `loanbook` must be unique.
      Do you need to remove duplicated values (e.g. `1`)?"
**Suggest edit** 
"All values of `id_loan` in a `loanbook` must be unique.
      Please ensure that every loan has a unique ID" 

**The reason being** - "remove Duplicate" could prompt a user to delete these rows. - These rows may have non-unique IDs however they may, in fact, unique loans. 
I.e. two loans to one company with the same ID.  (most likely due to an error on the bank's data entry side)
So we want to preserve both these loans and Identify them as being unique. 
i.e. two loans to one company with a unique ID each 

Thanks a lot !
@jdhoffa FYI

**Comment by maurolepore** [comment]
Thanks @georgeharris2deg, I changed the error message a bit to more closely reflect what 
you suggest. I still would like to preserve the hint about the position of the first duplicated 
value (in the example below, the first duplicated value is `1`).

``` r
devtools::load_all()
#> Loading r2dii.match
duplicated <- rep.int(1, times = 2)
lbk <- fake_lbk(id_loan = duplicated)
ald <- fake_ald()

try(match_name(lbk, ald))
#> Error :   All values of `id_loan` in a `loanbook` must be unique (`1` is not).
#>   Please ensure that every loan has a unique identifier.
```

<sup>Created on 2021-02-11 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

--

@jdhoffa, I moved this to a draft because I see some snapshots are now out of date and I need to figure out what's going on.

**Comment by maurolepore** [comment]
@jdhoffa, following the meeting we had last Friday I understand it's a priority to submit #353 to CRAN. I think it would be nice for that release to include this PR too -- as it was requested by @georgeharris2deg a while ago. 

I assume you don't have time to review so I'll go ahead and merge so I can start preparing the release. If you disagree, feel free to revert.

**Comment by jdhoffa** [comment]
This looks good! Apologies for slacking on my review duties.

**Pull request merged**